Improved zpool status output, list all affected datasets

gamanakis opened the pull request:
<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

<!--- Provide a general summary of your changes in the Title above -->

<!---
Documentation on ZFS Buildbot options can be found at
https://openzfs.github.io/openzfs-docs/Developer%20Resources/Buildbot%20Options.html
-->

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
This is a new effort to merge #9175 in current master.


### Description
<!--- Describe your changes in detail -->
Currently, determining which datasets are affected by corruption is
a manual process.

The primary difficulty in reporting the list of affected snapshots is
that since the error was initially found, the snapshot where the error
originally occurred in, may have been deleted. To solve this issue, we
add the ID of the head dataset of the original snapshot which the error
was detected in, to the stored error report. Then any time a filesystem
is deleted, the errors associated with it are deleted as well. Any time
a clone promote occurs, we modify reports associated with the original
head to refer to the new head. The stored error reports are identified
by this head ID, the birth time of the block which the error occurred
in, as well as some information about the error itself are also stored.

Once this information is stored, we can find the set of datasets
affected by an error by walking back the list of snapshots in the given
head until we find one with the appropriate birth txg, and then traverse
through the snapshots of the clone family, terminating a branch if the
block was replaced in a given snapshot. Then we report this information
back to libzfs, and to the zpool status command, where it is displayed
as follows:

 pool: test
 state: ONLINE
status: One or more devices has experienced an error resulting in data
        corruption.  Applications may be affected.
action: Restore the file in question if possible.  Otherwise restore the
        entire pool from backup.
   see: https://openzfs.github.io/openzfs-docs/msg/ZFS-8000-8A
  scan: scrub repaired 0B in 00:00:00 with 800 errors on Fri Dec  3
08:27:57 2021
config:

        NAME        STATE     READ WRITE CKSUM
        test        ONLINE       0     0     0
          sdb       ONLINE       0     0 1.58K

errors: Permanent errors have been detected in the following files:

        test@1:/test.0.0
        /test/test.0.0
        /test/1clone/test.0.0

A new feature flag is introduced to mark the presence of this change, as
well as promotion and backwards compatibility logic. This is an updated
version of #9175. Rebase required fixing the tests, updating the ABI of
libzfs, updating the man pages, fixing bugs, and updating the old
on-disk error logs to the new format upon activation of the feature.

Signed-off-by: TulsiJain <tulsi.jain@delphix.com>
Signed-off-by: George Amanakis <gamanakis@gmail.com>


### How Has This Been Tested?
Compiles and runs locally, the two introduced tests (zpool_status_00{3,4}) pass.
ZTS does not report any significant errors.
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If your change is a performance enhancement, please provide benchmarks here. -->
<!--- Please think about using the draft PR feature if appropriate -->

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [x] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [x] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [x] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

alek-p commented:
Thanks for working on this! Looks like it will also help with picking candidate snapshots for [healing recv](https://github.com/openzfs/zfs/pull/9372).

gamanakis commented:
Marked this as ready for review. I incorporated the suggestions by @allanjude from #9175.

tonyhutter commented on module/zcommon/zfeature_common.c:
```diff
@@ -594,6 +594,11 @@ zpool_feature_init(void)
+
+	zfeature_register(SPA_FEATURE_HEAD_ERRLOG,
+	    "com.delphix:head_errlog", "head_errlog",
+	    "Intoduce an on-disk block error log.",
```
"Introduce"

tonyhutter commented on module/zfs/spa_errlog.c:
```diff
@@ -67,9 +77,20 @@ bookmark_to_name(zbookmark_phys_t *zb, char *buf, size_t len)
 
 /*
- * Convert a string to a bookmark
+ * Convert a err_phys to a string.
```
a -> an

tonyhutter commented:
Your commit message is currently just:
```
Improved zpool status output, list all affected datasets
Signed-off-by: TulsiJain <tulsi.jain@delphix.com>
Signed-off-by: George Amanakis <gamanakis@gmail.com>
```
Please add more detail to the commit message.  At a minimum, include an example of what the new zpool status -v output looks like.  You should also remove the old signed-off-by lines, and add a signed-off-line with your name.  Also add a line saying:  "This is an updated version of #9175".

tonyhutter commented:
You should probably also mention in your commit message that this adds a feature flag.

tonyhutter commented on module/zfs/spa_errlog.c:
```diff
@@ -53,7 +53,17 @@
+#include <sys/dmu_objset.h>
+#include <sys/dbuf.h>
 
+#ifdef _KERNEL
```
I'm probably missing something obvious, but why is this `#ifdef   _KERNEL`?

gamanakis replied:
I am not sure either, it doesn't make much sense so I removed the `#ifdef`.

gamanakis replied:
It needs them, otherwise when compiling libzfs it fails with:
```
./../module/zfs/spa_errlog.c:60:12: error: ‘check_filesystem’ declared ‘static’ but never defined [-Werror=unused-function]
 static int check_filesystem(spa_t *spa, uint64_t head_ds,
            ^~~~~~~~~~~~~~~~
../../module/zfs/spa_errlog.c:63:17: error: ‘get_errlog_size’ declared ‘static’ but never defined [-Werror=unused-function]
 static uint64_t get_errlog_size(spa_t *spa, uint64_t spa_err_obj);
                 ^~~~~~~~~~~~~~~
../../module/zfs/spa_errlog.c:64:17: error: ‘get_errlist_size’ declared ‘static’ but never defined [-Werror=unused-function]
 static uint64_t get_errlist_size(spa_t *spa, avl_tree_t *tree);
                 ^~~~~~~~~~~~~~~~
```

allanjude replied:
You might be able to avoid needing these prototypes, if you just move the `check_filesystem()` function to before where it is called in `check_clones()` and `process_error_block` etc

gamanakis replied:
Nice suggestion, I moved them around and got rid of the prototypes.

tonyhutter commented:
Hmm... am I missing somewhere where you disable headerrlog for zpool_status_004_pos.ksh?  These test look nearly identical:

```
$ diff -u tests/zfs-tests/tests/functional/cli_root/zpool_status/zpool_status_003_pos.ksh tests/zfs-tests/tests/functional/cli_root/zpool_status/zpool_status_004_pos.ksh
--- tests/zfs-tests/tests/functional/cli_root/zpool_status/zpool_status_003_pos.ksh	2021-12-02 15:46:44.893262446 -0800
+++ tests/zfs-tests/tests/functional/cli_root/zpool_status/zpool_status_004_pos.ksh	2021-12-02 15:46:44.893262446 -0800
@@ -28,7 +28,7 @@
 
 #
 # DESCRIPTION:
-# Verify correct output with 'zpool status -v' after corrupting a file
+# Verify feature@headerrlog=disabled works perfectly.
 #
 # STRATEGY:
 # 1. Create a file system
@@ -66,4 +66,4 @@
 log_must eval "zpool status -v | grep '$TESTPOOL/testclone/10m_file'"
 log_must eval "zpool status -v | grep '$TESTDIR/10m_file'"
 
-log_pass "'zpool status -v' outputs affected filesystem, snapshot & clone"
+log_pass "'feature@headerrlog=disabled works perfectly"
```

gamanakis commented:
@tonyhutter thank you for helping out with some rough edges I missed, these are now fixed.

allanjude commented on module/zfs/spa_errlog.c:
```diff
@@ -82,8 +103,91 @@ name_to_bookmark(char *buf, zbookmark_phys_t *zb)
+		dsl_pool_config_exit(dp, FTAG);
+		return (EFAULT);
+	}
+	*head_dataset_id = dsl_dir_phys(ds->ds_dir)->dd_head_dataset_obj;
```
should we assert that head_dataset_id is not null, before dereferencing it?

gamanakis replied:
It doesn't show here in the conversation, but I added the assertion.

allanjude commented on module/zfs/spa_errlog.c:
```diff
@@ -136,83 +240,440 @@ spa_log_error(spa_t *spa, const zbookmark_phys_t *zb)
+		mutex_enter(&spa->spa_errlist_lock);
+		total += get_errlist_size(spa, &spa->spa_errlist_last);
+		total += get_errlist_size(spa, &spa->spa_errlist_scrub);
+		mutex_exit(&spa->spa_errlist_lock);
```
is there a reason to use a different order for the errlog than the errlist?

allanjude commented on module/zfs/spa_errlog.c:
```diff
@@ -136,83 +240,440 @@ spa_log_error(spa_t *spa, const zbookmark_phys_t *zb)
+{
+	objset_t *os;
+	if (dmu_objset_from_ds(ds, &os) != 0) {
+		return (ENOENT);
```
the returns in this function should have SET_ERROR() wrappers

allanjude commented on module/zfs/spa_errlog.c:
```diff
@@ -136,83 +240,440 @@ spa_log_error(spa_t *spa, const zbookmark_phys_t *zb)
+		if (err != 0) {
 			zap_cursor_fini(&zc);
-			return (SET_ERROR(ENOMEM));
+			return (EFAULT);
```
why drop the SET_ERROR()?

tonyhutter commented on module/zfs/spa_errlog.c:
```diff
@@ -128,6 +225,289 @@ spa_log_error(spa_t *spa, const zbookmark_phys_t *zb)
+	dnode_rele(dn, FTAG);
+	return (0);
+}
+
```
This is a pretty big function.  Can you add a comment describing this it does, what its arguments are, and what the return value is?

tonyhutter commented on module/zfs/spa_errlog.c:
```diff
@@ -404,6 +900,134 @@ spa_errlog_sync(spa_t *spa, uint64_t txg)
+	mutex_exit(&spa->spa_errlog_lock);
+}
+
+#if defined(_KERNEL)
```
Is `#if defined(_KERNEL)` needed here?

gamanakis replied:
`swap_errlog()` references `name_to_errphys()` which is also in an `ifdef _KERNEL` above.

gamanakis replied:
I could move `name_to_errphys()` above out of its `ifdef _KERNEL` block, and that would allow removing those ifdefs.

tonyhutter commented on module/zfs/spa_errlog.c:
```diff
@@ -404,6 +900,134 @@ spa_errlog_sync(spa_t *spa, uint64_t txg)
+    dmu_tx_t *tx)
+{
+	mutex_enter(&spa->spa_errlog_lock);
+#if defined(_KERNEL)
```
Is `#if defined(_KERNEL)` needed here?

gamanakis replied:
See above.

tonyhutter commented on tests/zfs-tests/tests/functional/cli_root/zpool_status/zpool_status_004_pos.ksh:
```diff
@@ -0,0 +1,81 @@
+log_mustnot eval "zpool status -v | grep '$TESTPOOL/testclone/10m_file'"
+log_must eval "zpool status -v | grep '$TESTDIR/10m_file'"
+
+log_pass "'feature@head_errlog=disabled works"
```
suggestion:
 ```
log_pass "'zpool status -v' with feature@head_errlog=disabled works" 
```

gamanakis commented:
I also need to fix releasing the hold of a dataset in case of an error in check_filesystem().

gamanakis commented:
04296bf: Addressed suggestions, added comments, fixed some typos and a subtle bug in check_filesystem() in case of an error.

tonyhutter commented on module/zfs/spa_errlog.c:
```diff
@@ -128,6 +225,296 @@ spa_log_error(spa_t *spa, const zbookmark_phys_t *zb)
+ * This function serves a double role. If only_count is true, it returns
+ * how many times an error block belonging to this filesystem is referenced by
+ * snapshots or clones. If only_count is false, each time the error block is
+ * referenced by a snapshot or clone, it returns the address of zep back to
```
zep -> zap

gamanakis commented:
0fe047a: Fixed a comment in check_filesystem().

tonyhutter reviewed (approved):
Thanks for sticking with this.  I think this is going to be a very useful feature!

behlendorf commented on module/zfs/spa_errlog.c:
```diff
@@ -304,28 +760,75 @@ sync_error_list(spa_t *spa, avl_tree_t *t, uint64_t *obj, dmu_tx_t *tx)
+			int error = zap_lookup_int_key(spa->spa_meta_objset,
+			    *obj, head_dataset_obj, &err_obj);
+
+			if (error != 0) {
```
Only on `ENOENT` would we want to create the error log.

gamanakis commented on man/man7/zpool-features.7:
```diff
@@ -832,7 +832,15 @@ once all filesystems that have ever had their
+With this feature enabled, every dataset affected by an error block is listed
+in the output of
+.Nm zpool Cm status .
+This feature can only be enabled at pool creation time.
```
Nice suggestion, let me see what I can do.

gamanakis replied:
Introduced a new function `update_errlog()` to update the old error logs to the new on-disk format.

gamanakis commented:
Happy new year to all!

I figured out how to reset the old error log and activate the feature without needing to recreate the pool from scratch. Will look into how to convert the old error log to the new format (and also report errors in underlying snapshots/clones).

gamanakis commented on include/sys/zio.h:
```diff
@@ -291,6 +291,13 @@ struct zbookmark_phys {
 	uint64_t	zb_blkid;
 };
 
+typedef struct zbookmark_err_phys {
```
Done.

gamanakis commented on tests/zfs-tests/tests/functional/cli_root/zpool_status/zpool_status_004_pos.ksh:
```diff
@@ -0,0 +1,81 @@
+log_assert "Verify 'zpool status -v' with feature@head_errlog=disabled works"
+log_onexit cleanup
+
+log_must zpool create -f -o feature@head_errlog=disabled $TESTPOOL $DISKS
```
Fixed.

gamanakis commented on module/zfs/spa_errlog.c:
```diff
@@ -53,7 +53,9 @@
 #include <sys/zap.h>
 #include <sys/zio.h>
-
+#include <sys/dsl_dir.h>
```
Done.

gamanakis commented:
ce3afa6: addressed suggestions from @behlendorf, updated comments and man pages. Most importantly wrote a new function `update_errlog()` which updates the on-disk error logs to the new format upon activation of the new feature.

nabijaczleweli commented on man/man7/zpool-features.7:
```diff
@@ -832,7 +832,14 @@ once all filesystems that have ever had their
+in the output of
+.Nm zpool Cm status .
 .El
-.
```
Why remove this?

gamanakis replied:
Thanks for catching it, will fix.

behlendorf commented on module/zfs/spa_errlog.c:
```diff
@@ -148,71 +596,167 @@ spa_get_errlog_size(spa_t *spa)
+	 * error logs by spa_errlog_sync().
+	 */
+	for (uint64_t i = 0; i < alloc; i++)
+		spa_log_error(spa, &zb[i]);
```
Perhaps we could simplify this with a small bit of code duplication.  Rather than relying on `spa_log_error()` what if the code iterated over both the `scrub` and `last` error logs, wrote out the new objects, then removed the old log objects and referenced the new ones?

I think this would have a couple of advantages:

1) By keeping the log conversion and feature activation in a single txg we're guaranteed never to be in an inconsistent state if the system crashes.  Either the feature is disabled and we have the old log format, or it's enabled and everything on disk has been converted.
2) No need for a potentially large memory allocation for all the records.  While unlikely, the allocation may be too large if the pool had millions of reported errors.  
3) In the event that the pool does have a huge number of errors, we could opt to only convert say the first 1000 and discard the rest.  That's not ideal, but it's probably preferable to having a huge txg sync time (which would look like a hang) while all of the block birth txgs are looked up.  In this unlikely worst case, the next scrub would add the records which weren't converted which isn't so bad.

Partial untested pseudo code for the basic idea.

```c
        uint64_t obj = zap_create(spa->spa_meta_objset, DMU_OT_ERROR_LOG,
            DMU_OT_NONE, 0, tx);

        for (zap_cursor_init(&zc, spa->spa_meta_objset, spa->spa_errlog_scrub);
            zap_cursor_retrieve(&zc, &za) == 0;
            zap_cursor_advance(&zc)) {
                name_to_bookmark(za.za_name, &zb);

                zbookmark_err_phys_t zep;
                zep.zb_object = zb.zb_object;
                zep.zb_level = zb.zb_level;
                zep.zb_blkid = zb.zb_blkid;

                uint64_t head_dataset_obj;
                get_head_and_birth_txg(spa, &zep,
                    zb.zb_objset, &head_dataset_obj);

                uint64_t err_obj;
                int error = zap_lookup_int_key(spa->spa_meta_objset,
                    obj, head_dataset_obj, &err_obj);

                if (error == ENOENT) {
                        err_obj = zap_create(spa->spa_meta_objset,
                            DMU_OT_ERROR_LOG, DMU_OT_NONE, 0, tx);

                        (void) zap_update_int_key(spa->spa_meta_objset,
                            obj, head_dataset_obj, err_obj, tx);
                }

                char buf[64];
                errphys_to_name(&zep, buf, sizeof (buf));

                (void) zap_update(spa->spa_meta_objset,
                    err_obj, buf, 1, strlen(name) + 1, name, tx);
        }
```

gamanakis replied:
Makes sense, let me see how I can implement this.

behlendorf commented on module/zfs/spa_errlog.c:
```diff
@@ -136,9 +533,60 @@ spa_log_error(spa_t *spa, const zbookmark_phys_t *zb)
+ * log format.
+ */
+void
+update_errlog(spa_t *spa, dmu_tx_t *tx)
```
```suggestion
sync_upgrade_errlog(spa_t *spa, dmu_tx_t *tx)
```

behlendorf commented on module/zfs/spa_errlog.c:
```diff
@@ -136,9 +533,60 @@ spa_log_error(spa_t *spa, const zbookmark_phys_t *zb)
+	return (total);
+}
+
+#ifdef _KERNEL
```
We should enable all of this code in both the user and kernel builds do it's available as needed for tools like `ztest` and `zhack`.  Unless that's not possible for some reason?

behlendorf reviewed (commented):
Thanks for tackling the upgrade logic, I think getting rid of that restriction makes this a lot more useful.  Generally this looks good, just a couple comments about the upgrade strategy.

gamanakis commented:
e96984f: Rewrote the upgrade logic according to the suggestion of @behlendorf. Also introduced a tunable (`spa_upgrade_errlog_limit`) to control the number of error entries that will be upgraded to the new on-disk format.

nabijaczleweli commented on module/zfs/spa_errlog.c:
```diff
@@ -53,7 +63,17 @@
+ * 		format when enabling head_errlog. Defaults to 0 which ignores
+ * 		the setting.
+ */
+uint32_t spa_upgrade_errlog_limit = 0;
```
`static`

gamanakis replied:
I have the impression that most other tunables are not defined as static, although they are used only in the file where they are defined.

nabijaczleweli replied:
#12899; I don't necessarily want to spend even more hours running that on random new code.

gamanakis replied:
Great! Had not seen that. Will fix here.

gamanakis commented:
a742722: minor fixes, removing an `_if kernel` and declaring the tunable as static.

behlendorf commented on module/zfs/spa_errlog.c:
```diff
@@ -128,6 +243,296 @@ spa_log_error(spa_t *spa, const zbookmark_phys_t *zb)
+
+		dsl_dataset_rele(ds, FTAG);
+		if (dsl_dataset_hold_obj(dp, snap_obj, FTAG, &ds) != 0)
+			return (SET_ERROR(EFAULT));
```
`snap_obj_array` is leaked here.

behlendorf commented on module/zfs/spa_errlog.c:
```diff
@@ -128,6 +243,296 @@ spa_log_error(spa_t *spa, const zbookmark_phys_t *zb)
+	return (0);
+out:
+	kmem_free(snap_obj_array, sizeof (*snap_obj_array));
+	return (SET_ERROR(EFAULT));
```
How about a single exit path which returns `err`, which can be set before the `goto`.

```
out:
	kmem_free(snap_obj_array, sizeof (*snap_obj_array));
	return (err);
``

behlendorf commented on module/zfs/spa_errlog.c:
```diff
@@ -128,6 +243,296 @@ spa_log_error(spa_t *spa, const zbookmark_phys_t *zb)
+				zbookmark_phys_t zb;
+				zep_to_zb(snap_obj, zep, &zb);
+				if (copyout(&zb, (char *)addr + (*count - 1)
+				    * sizeof (zbookmark_phys_t),
```
cstyle: The `*` should be at the end of the previous line.

behlendorf reviewed (commented):
Thanks for the quick turn around.  I've posted another round of feedback, but I didn't quite make it through every function.  Locally it worked quite well for me which was great.

gamanakis commented:
Changed the error handling of all functions to return the actual error. However I broke things probably in the `zfs_ioc_pool_stats()->spa_get_stats()->spa_get_errlog_size()` path, leading to lockups. Will take another look.

gamanakis commented:
e8578f3: Fixed the error return codes and added comments. Local testing revealed no lockups anymore.

behlendorf commented:
It looks like my suggested changes to the upgrade path introduced an issue though, http://build.zfsonlinux.org/builders/Fedora%2035%20x86_64%20%28TEST%29/builds/328/steps/shell_4/logs/console

gamanakis commented:
@behlendorf right, I just saw this too. Let me see what is going on.

gamanakis commented:
I know what is going on: I introduced a `spa_log_error()` in line 972 of `spa_errlog.c` in case `get_head_and_birth_txg()` returns an error. I though it would be nice to re-insert the error block into the error list, apparently that led to the panic. 

We could also opt out and let it continue by loosing the error block (ie not writing it to the on-disk error log, since `get_head_and_birth_txg() returns an error. This was the behavior in the original code.

gamanakis commented:
9768d24: Fixed the previous panic, which unmasked that `zpool_events_duplicates` is reporting duplicates of errors and failing.

gamanakis commented:
67583f8: fixed the failure in `zpool_events_duplicates.ksh`: `NOTE: UNEXPECTED -- 12 duplicate io write ereports`. It seems we should simply discard error blocks if get_head_and_birth_txg() fails in sync_error_list(). Otherwise if we reinsert them in the error lists (even if we are not syncing the final txg) will result in duplicate posting of errors.

tonyhutter commented on module/zfs/spa_errlog.c:
```diff
@@ -128,91 +249,600 @@ spa_log_error(spa_t *spa, const zbookmark_phys_t *zb)
+		spa->spa_errlog_last = newobj;
+	}
+
+	if (spa->spa_errlog_scrub != 0) {
+		sync_upgrade_errlog(spa, spa->spa_errlog_scrub, &newobj, tx);
+		mutex_exit(&spa->spa_errlog_lock);
+		spa->spa_errlog_scrub = newobj;
+	}
+	mutex_exit(&spa->spa_errlog_lock);
```
It looks like you're calling `mutex_exit(&spa->spa_errlog_lock)` twice if `(spa->spa_errlog_scrub != 0)`.  Is that expected?

gamanakis replied:
Good catch, will fix.

tonyhutter replied:
I think this was my only change - give me a heads-up when this is in and I'll approve it.

gamanakis replied:
Done, thank you!

ahrens commented:
@gamanakis Thanks a lot for picking this up!  Someone from Delphix should review this, but we might not have time for an in-depth review until February.

gamanakis commented:
b205135: rebased to master, removed double mutex_exit().

tonyhutter commented on module/zfs/spa_errlog.c:
```diff
@@ -128,91 +249,599 @@ spa_log_error(spa_t *spa, const zbookmark_phys_t *zb)
+ * sum of both the last log and the current log, since we don't know the union
+ * of these logs until we reach userland.
+ */
+void
+spa_get_errlog_size(spa_t *spa, uint64_t *count)
```
```diff
- uint64_t spa_get_errlog_size(spa_t *spa)
+ void spa_get_errlog_size(spa_t *spa, uint64_t *count)
```
In every place `spa_get_errlog_size()` is called, `count` is always zeroed first, so there doesn't seem to be a reason for this change.

gamanakis replied:
Correct, I will simplify this to directly return the `count`.

gamanakis commented:
843d1a9: rebased to master, restore `spa_get_errlog_size()` to directly return the size (instead of using a pointer).

tonyhutter reviewed (approved):
Thanks @gamanakis

gamanakis commented:
[a8a44b9](https://github.com/openzfs/zfs/commit/a8a44b909e63b7d6e8e2d45f34117d6ba53ddae8): rebased to master.

gamanakis commented:
[44496bd](https://github.com/openzfs/zfs/pull/12812/commits/44496bdf95eedcb1fdb6240309fe112a356bf81b): Rebased to master.

behlendorf commented:
@ahrens it'd be great if someone from Delphix could take a look at this in the next few weeks.

@gamanakis sorry, but it looks like another rebase is going to be needed.

gamanakis commented:
[4105d2b](https://github.com/openzfs/zfs/commit/4105d2ba2db4fc31cdda8fb8c4438fb643586ad1): Rebased to master.

gamanakis commented:
I need to take a deeper look at this, the newly introduced tests are failing in the CI after the rebase.

gamanakis commented:
I figured what is going on: Since enabling compression by default (56fa4aa96eb3875f254e93eaef646ea20ba187f9), `zinject` cannot taint files that are fully compressed, such as `0` filled files that `mkfile` creates. I will switch to using `fio`.

gamanakis commented:
79a219f : Fix the tests.

nabijaczleweli commented on man/man7/zpool-features.7:
```diff
@@ -507,6 +507,15 @@ once either of the limit properties has been set on a dataset
+and keyed by the head id.
+With this feature enabled, every dataset affected by an error block is listed
+in the output of
+.Nm zpool Cm status .
```
This could also use one of the standard state machine spiels, I think `\*[instant-never]` for this (but dunno)?

gamanakis replied:
I didn't understand your comment, could you elaborate?

nabijaczleweli replied:
All (most? but I tried to make it all) other features have a short paragraph at the end describing the disabled<->enabled<->active state machine – for example `embedded_data` has `\*[instant-never]` (which expands to `.No This feature becomes Sy active No as soon as it is enabled and will never return to being Sy enabled .`), checksum features use `.checksum-spiel SUM` (which expands to `.No When the .Sy SUM No feature is set to Sy enabled , No the administrator can turn on the Sy SUM No checksum on any ...` &c.), `large_blocks` has a different set of conditions (`No This feature becomes Sy active No once a dataset contains a file with a block size larger than 128kB, and will return to being Sy enabled No once all ...`).

If this behaves as I think it does – i.e. like `embedded_data`, in that it activates instantly and you can never turn it off – then all you need to do here is add
```
.Nm zpool Cm status .
.Pp
\*[instant-never]
.
.feature ...
```
Otherwise copy-and-modify one of the other descriptions.

gamanakis replied:
Thank you, fixed in the upcoming rebase.

gamanakis commented:
[e51ec30](https://github.com/openzfs/zfs/commit/e51ec30c0e4ed0c81bbc7e331c5b53f7a6718dcc): Rebase to master.

ahrens commented:
@gamanakis @behlendorf It's great to see this, thanks for picking it up!  I or someone else from Delphix will do a code review in the next couple of weeks.

behlendorf commented:
@ahrens @mmaybee it'd still be great to get some eyes from Delphix on this change.

@gamanakis another rebase would probably be a good idea.  There's been a fair bit of change to the master branch in the last month.

gamanakis commented on module/zfs/spa_errlog.c:
```diff
@@ -82,8 +129,82 @@ name_to_bookmark(char *buf, zbookmark_phys_t *zb)
+	if (error == 0 && BP_IS_HOLE(&bp))
+		error = SET_ERROR(ENOENT);
+
+	if (error != 0) {
+		rw_exit(&dn->dn_struct_rwlock);
+		dnode_rele(dn, FTAG);
+		dsl_dataset_rele(ds, FTAG);
+		dsl_pool_config_exit(dp, FTAG);
+		return (error);
+	}
+
+	zep->zb_birth = bp.blk_birth;
+	rw_exit(&dn->dn_struct_rwlock);
+	dnode_rele(dn, FTAG);
+	dsl_dataset_rele(ds, FTAG);
+	dsl_pool_config_exit(dp, FTAG);
+	return (0);
```
Makes sense, it should be OK to return "garbage" in zb_birth.

gamanakis commented on module/zfs/spa_errlog.c:
```diff
@@ -128,6 +249,317 @@ spa_log_error(spa_t *spa, const zbookmark_phys_t *zb)
+			    &total, NULL, B_TRUE) != 0) {
+				zap_cursor_fini(&head_ds_cursor);
+				zap_cursor_fini(&zc);
+				return (total);
```
`get_errlog_size()` is called by `spa_get_errlog_size()`, and both do not have the ability to return an error, meaning in case of errors we will end up with a partial total.

I think your suggestion to "drive on" by advancing the cursor makes more sense here as the total reported will be closer to the "real" value, had an error not occurred.

mmaybee replied:
OK. Are there any negative consequences of an incorrect total here? Where is this used? Is it just part of the reported output?

gamanakis replied:
I do not think there are any negative sequences of a partial total. All the upstream functions calling `spa_get_errlog_size()` return output to the user.

`spa_get_errlog_size()` is used:
1. in  `zpool_refresh_stats()` -> `zfs_ioc_pool_stats()` -> `spa_get_stats()`, ZPOOL_CONFIG_ERRCOUNT is populated. This is used for userspace reporting in `zpool_get_errlog()`, `zpool_do_status()` and `check_status()`.
2. in `zpool_get_errlog()` -> `zfs_ioc_error_log()` when `spa_get_errlog()` returns an error, it turns to `spa_get_errlog_size()` to determine the size of the error logs.
3. in `dsl_scan_done()` when logging in spa history

mmaybee replied:
OK, great

gamanakis commented on module/zfs/spa_errlog.c:
```diff
@@ -128,6 +249,317 @@ spa_log_error(spa_t *spa, const zbookmark_phys_t *zb)
+		uint64_t head_ds_obj;
+		int error = get_head_and_birth_txg(spa, &zep,
+		    se->se_bookmark.zb_objset, &head_ds_obj);
+		if (error != 0)
+			continue;
+
+		(void) process_error_block(spa, head_ds_obj, &zep, &total,
+		    NULL, B_TRUE);
```
I think this is the more appropriate semantic (see comment above).

gamanakis commented on module/zfs/spa_errlog.c:
```diff
@@ -138,81 +570,279 @@ spa_get_errlog_size(spa_t *spa)
+
+		/*
+		 * We cannot use get_head_and_birth_txg() because it will
+		 * acquire the pool config lock, which we already have. In case
+		 * of an error we simply continue.
```
Yes, if `sync_upgrade_errlog()` cannot figure out the dataset, the objset, and the dnode of the object then it will not upgrade the old error entry and will discard it at the end.

This may happen in two cases:
1) The object does not actually belong to any dataset anymore (eg. a file was deleted and is not present in any snapshot or clone).
2) The object belongs to a dataset but `sync_upgrade_errlog()` cannot figure that out (because `dsl_dataset_hold_obj()`, `dmu_objset_from_ds()` or `dnode_hold()` error out).

For the zap entry we actually need to know only the dataset the block belongs to. The objset and the dnode are used to determine the block pointer and then if `BP_IS_HOLE()`. If we fail to retrieve the objset and the dnode (and subsequently fail to determine `BP_IS_HOLE()`) would it be reasonable to err on the side of caution and still create a zap entry (presuming that we know the dataset)?

Do you have another suggestion?

mmaybee replied:
Hm, good question. I lean toward the current semantic of dropping the error if we can't "verify" it here, otherwise we might get spurious errors that we can't ever resolve? An error here could indicate further corruption/errors and if this error is truly persistent, it should re-appear after a scan.

gamanakis replied:
Seems reasonable to me, I will leave it as it is and expand the comment.

gamanakis commented:
@mmaybee thank you for taking a look. Most of the comments are pretty straightforward and I already implemented them in my branch (marked as resolved). In a couple of others I provided responses.

gamanakis commented on module/zfs/spa_errlog.c:
```diff
@@ -406,6 +1091,136 @@ spa_errlog_sync(spa_t *spa, uint64_t txg)
+	uint64_t snap_obj = dsl_dataset_phys(ds)->ds_prev_snap_obj;
+	uint64_t snap_obj_txg = dsl_dataset_phys(ds)->ds_prev_snap_txg;
+
+	while (snap_obj != 0) {
+		dsl_dataset_rele(ds, FTAG);
+		if (dsl_dataset_hold_obj(dp, snap_obj, FTAG, &ds) != 0)
+			continue;
```
Good catch, thank you. We should exit with an error if the call to `dsl_dataset_hold_obj()` fails.

gamanakis commented on module/zfs/spa_errlog.c:
```diff
@@ -128,6 +249,317 @@ spa_log_error(spa_t *spa, const zbookmark_phys_t *zb)
+				    sizeof (zbookmark_phys_t),
+				    sizeof (zbookmark_phys_t)) != 0) {
+					dsl_dataset_rele(ds, FTAG);
+					goto out;
```
Thank you, done.

gamanakis commented:
[e09ae05](https://github.com/openzfs/zfs/pull/12812/commits/e09ae05789c808f5ee84f93142446023a2d1fdc7): added a comment in `sync_upgrade_errlog()`

ahrens commented on module/zfs/spa_errlog.c:
```diff
@@ -128,6 +241,300 @@ spa_log_error(spa_t *spa, const zbookmark_phys_t *zb)
+ * how many times an error block belonging to this filesystem is referenced by
+ * snapshots or clones. If only_count is false, each time the error block is
+ * referenced by a snapshot or clone, it returns a pointer to the bookmark of
+ * the error block back to process_error_log().
```
I don't see a pointer being returned.  It looks like are filling the userspace array at `addr` with `zbookmark_phys_t`'s.  The array has space for `*count` entries (each entry holds a `zbookmark_phys_t`).  The array is filled from the back, and the *count is modified to be the number of unused entries at the beginning of the array.

This seems like an odd interface.  Maybe it's following an existing precedent?  In which case we should at least document it well.  If not, could we instead fill the array from the beginning, and return the number of entries added?

gamanakis replied:
I realized my wording is wrong. It does not return a pointer, it copies to userspace (in *addr, or *uaddr if you prefer) the bookmark of the error block. This is a pre-existing interface, it was not implemented in this PR.

ahrens commented on module/zfs/spa_errlog.c:
```diff
@@ -128,6 +241,300 @@ spa_log_error(spa_t *spa, const zbookmark_phys_t *zb)
+
+/*
+ * This function serves a double role. If only_count is true, it returns
+ * how many times an error block belonging to this filesystem is referenced by
```
How about `it returns (in *count) how many times ...`.  Since the actual return value is not the count but rather an error code.

ahrens commented on module/zfs/spa_errlog.c:
```diff
@@ -136,83 +543,284 @@ spa_log_error(spa_t *spa, const zbookmark_phys_t *zb)
 
 #ifdef _KERNEL
+/*
+ * If an error block is shared by two datasets it would be counted twice. For
```
`would` -> `will` ?

ahrens commented on module/zfs/spa_errlog.c:
```diff
@@ -136,83 +543,284 @@ spa_log_error(spa_t *spa, const zbookmark_phys_t *zb)
+ */
 static int
-process_error_log(spa_t *spa, uint64_t obj, void *addr, size_t *count)
+process_error_log(spa_t *spa, uint64_t obj, void *addr, uint64_t *count)
```
To give a clue that `addr` is a userspace pointer, maybe we can name it `uaddr`, as is done in spa_get_errlog().  (this applies to many functions in this file)

ahrens commented on module/zfs/spa_errlog.c:
```diff
@@ -136,83 +543,284 @@ spa_log_error(spa_t *spa, const zbookmark_phys_t *zb)
-			return (SET_ERROR(EFAULT));
+		zap_cursor_t head_ds_cursor;
+		zap_attribute_t head_ds_attr;
+		zbookmark_err_phys_t head_ds_block;
```
I think the declaration of `head_ds_block` can be moved inside the loop below (just before it's passed to name_to_errphys()).

ahrens commented on module/zfs/spa_errlog.c:
```diff
@@ -128,6 +241,300 @@ spa_log_error(spa_t *spa, const zbookmark_phys_t *zb)
+		return (error);
+
+	uint64_t snap_obj = dsl_dataset_phys(ds)->ds_prev_snap_obj;
+	uint64_t snap_obj_txg = dsl_dataset_phys(ds)->ds_prev_snap_txg;
```
Since this is specifically the snapshot previous to `ds`, maybe name these `prev_obj` and `prev_obj_txg`

ahrens commented on module/zfs/spa_errlog.c:
```diff
@@ -128,6 +241,300 @@ spa_log_error(spa_t *spa, const zbookmark_phys_t *zb)
+
+	uint64_t snap_obj = dsl_dataset_phys(ds)->ds_prev_snap_obj;
+	uint64_t snap_obj_txg = dsl_dataset_phys(ds)->ds_prev_snap_txg;
+	*top_affected_fs = head_ds;
```
Why is this needed?  Do we need to fill it in even on error return (line 448)?  If so then don't we need to fill it in on line 438?

gamanakis replied:
Not needed, I will remove it.

ahrens commented on module/zfs/spa_errlog.c:
```diff
@@ -128,6 +241,300 @@ spa_log_error(spa_t *spa, const zbookmark_phys_t *zb)
+	uint64_t snap_obj_txg = dsl_dataset_phys(ds)->ds_prev_snap_txg;
+	*top_affected_fs = head_ds;
+
+	while (snap_obj != 0 && zep->zb_birth < snap_obj_txg) {
```
I think that this could walk past this dataset's origin (clone point).  But we might expect that to not happen since zb_birth shouldn't be less than the origin?  Could we add assertions to that effect?

gamanakis replied:
Seems reasonable, I was thinking the following: Figure out the `dd_head_dataset_obj` of the current filesystem, get its `ds_creation_txg`, and then assert that it is not greater than the current snapshot's creation txg. Does this make sense?

ahrens replied:
Given your explanation of this function elsewhere, it sounds like the entire point of it involves walking past the origin, to determine which head filesystem contains the first snapshot that contains this block.  So this assertion would be invalid?

gamanakis replied:
Yes, you are right! I got confused myself.

ahrens commented on module/zfs/spa_errlog.c:
```diff
@@ -128,6 +241,300 @@ spa_log_error(spa_t *spa, const zbookmark_phys_t *zb)
+
+	int aff_snap_count = 0;
+	uint64_t snap_obj = dsl_dataset_phys(ds)->ds_prev_snap_obj;
+	uint64_t snap_obj_txg = dsl_dataset_phys(ds)->ds_prev_snap_txg;
```
Since check_filesystem() is passed the *oldest* affected snapshot, I think that this should be walking *forward*, but it's walking backwards, so I'm confused.  Can you explain how this all works?

gamanakis replied:
See my comments above.

ahrens commented on module/zfs/spa_errlog.c:
```diff
@@ -128,6 +241,300 @@ spa_log_error(spa_t *spa, const zbookmark_phys_t *zb)
+	if (error == 0) {
+		if (zep->zb_birth < latest_txg) {
+			txg_to_consider = latest_txg;
+		} else {
```
in this case should zb_birth==latest_txg, i.e. the block in this snapshot is the same as the block that had the error?  Maybe the check should be 
```c
if (latest_txg == zep->zb_birth) {
    ... add to count
} else {
...
}

gamanakis commented on module/zfs/spa_errlog.c:
```diff
@@ -128,6 +241,300 @@ spa_log_error(spa_t *spa, const zbookmark_phys_t *zb)
+}
+
+static int
+find_top_affected_fs(spa_t *spa, uint64_t head_ds, zbookmark_err_phys_t *zep,
```
I will add the comments, and move this to `dsl_dataset.c`. However, I think this function returns the `dd_head_dataset_obj` of the oldest snapshot which I believe is different than the oldest snapshot. Please correct me if I am wrong.

ahrens replied:
I see.  It might make sense to have a simpler, more general function like I described above, and then in spa_errlog.c we can go from the oldest snapshot obj to its head dataset (even though this does involve one more call to dsl_dataset_hold_obj()).

gamanakis commented on module/zfs/spa_errlog.c:
```diff
@@ -128,6 +241,300 @@ spa_log_error(spa_t *spa, const zbookmark_phys_t *zb)
+ * the error block back to process_error_log().
+ */
+static int
+check_filesystem(spa_t *spa, uint64_t fs, zbookmark_err_phys_t *zep,
```
I think `fs` is a head filesystem. `find_top_affected_fs()` returns the `dd_head_dataset_obj` of the oldest snapshot, not the oldest snapshot per se. Please do correct me if I misunderstood this.

gamanakis commented:
@ahrens thank you for taking the time to look over this. Most of the comments are straightforward, I marked them as resolved and implemented them. In the remaining ones I added responses.

ahrens commented on module/zfs/spa_errlog.c:
```diff
@@ -128,6 +241,302 @@ spa_log_error(spa_t *spa, const zbookmark_phys_t *zb)
+			for (int i = 0; i < aff_snap_count && !found; i++) {
+				found =
+				    (dsl_dir_phys(clone->ds_dir)->dd_origin_obj
+				    == snap_obj_array[i]);
```
This section could use a comment explaining that only clones whose origins were affected could also have affected snapshots.  But it will be more clear if you use `ds_next_clones_obj`

gamanakis commented on module/zfs/spa_errlog.c:
```diff
@@ -128,6 +241,302 @@ spa_log_error(spa_t *spa, const zbookmark_phys_t *zb)
+		zap_cursor_t zc;
+		zap_attribute_t za;
+
+		for (zap_cursor_init(&zc, spa->spa_meta_objset, zap_clone);
```
Nice suggestion, thank you!

gamanakis commented:
5e7ae76: Implemented suggestions from @ahrens, rebased to master.

behlendorf commented:
Thanks everyone for sticking with this PR to get it over the finish line.
